# Poker hands
Source: https://projecteuler.net/problem=54

# Problem statement
In the card game poker, a hand consists of five cards and are ranked, from lowest to highest, in the following way:

High Card: Highest value card.
One Pair: Two cards of the same value.
Two Pairs: Two different pairs.
Three of a Kind: Three cards of the same value.
Straight: All cards are consecutive values.
Flush: All cards of the same suit.
Full House: Three of a kind and a pair.
Four of a Kind: Four cards of the same value.
Straight Flush: All cards are consecutive values of same suit.
Royal Flush: Ten, Jack, Queen, King, Ace, in same suit.
The cards are valued in the order:
2, 3, 4, 5, 6, 7, 8, 9, 10, Jack, Queen, King, Ace.

If two players have the same ranked hands then the rank made up of the highest value wins; for example, a pair of eights beats a pair of fives (see example 1 below). But if two ranks tie, for example, both players have a pair of queens, then highest cards in each hand are compared (see example 4 below); if the highest cards tie then the next highest cards are compared, and so on.

Consider the following five hands dealt to two players:

Hand	 	Player 1	 	Player 2	 	Winner

1	 	5H 5C 6S 7S KD
Pair of Fives
 	2C 3S 8S 8D TD
Pair of Eights
 	Player 2
    
2	 	5D 8C 9S JS AC
Highest card Ace
 	2C 5C 7D 8S QH
Highest card Queen
 	Player 1
    
3	 	2D 9C AS AH AC
Three Aces
 	3D 6D 7D TD QD
Flush with Diamonds
 	Player 2
    
4	 	4D 6S 9H QH QC
Pair of Queens
Highest card Nine
 	3D 6D 7H QD QS
Pair of Queens
Highest card Seven
 	Player 1
    
5	 	2H 2D 4C 4D 4S
Full House
With Three Fours
 	3C 3D 3S 9S 9D
Full House
with Three Threes
 	Player 1
    
The file, poker.txt, contains one-thousand random hands dealt to two players. Each line of the file contains ten cards (separated by a single space): the first five are Player 1's cards and the last five are Player 2's cards. You can assume that all hands are valid (no invalid characters or repeated cards), each player's hand is in no specific order, and in each hand there is a clear winner.

How many hands does Player 1 win?

# Planning
Create a function that reads in the data from the file and returns pairs of tuples/lists

Create a function that compares two hands
* Create functions that determine the value of each hand according to each different possible ranks
* Create a compareto function that compares 
    * Inspire by the java `compareTo` function
* 

In [60]:
def card_tuple_from_str(card_str):
    numeric_val = 0
    suite = ''

    if card_str[0] == 'T':
        numeric_val = 10
    elif card_str[0] == 'J':
        numeric_val = 11
    elif card_str[0] == 'Q':
        numeric_val = 12
    elif card_str[0] == 'K':
        numeric_val = 13
    elif card_str[0] == 'A':
        numeric_val = 14
    else:
        numeric_val = int(card_str[0])

    suite = card_str[1]

    card_tuple = ( numeric_val, suite )

    return card_tuple

In [61]:
card_tuple_from_str('JH')

(11, 'H')

In [4]:
## testing
card_tuple_from_str('4H')

(4, 'H')

In [5]:
## testing
card_tuple_from_str('AH')

(14, 'H')

In [2]:
def read_hand_string(line):
    
    line_divided = line.split()
    
    hand_1_str = line_divided[0:5]
    hand_1_tups = []
    for card_str in hand_1_str:
        hand_1_tups.append(card_tuple_from_str(card_str))
            
    
    hand_2_str = line_divided[5:10]
    hand_2_tups = []
    for card_str in hand_2_str:
        hand_2_tups.append(card_tuple_from_str(card_str))
      
    return hand_1_tups, hand_2_tups

In [6]:
## testing
read_hand_string('5H 5C 6S 7S KD 2C 3S 8S 8D TD')

([(5, 'H'), (5, 'C'), (6, 'S'), (7, 'S'), (0, 'D')],
 [(2, 'C'), (3, 'S'), (8, 'S'), (8, 'D'), (0, 'D')])

In [32]:
def high_card(hand_tups):
    return max([tup[0] for tup in hand_tups])
    
def one_pair(hand_tups):
    numeric_vals = [tup[0] for tup in hand_tups]
    
    highest_pair = 0
    
    for val in numeric_vals:
        if numeric_vals.count(val) >= 2:
            if val > highest_pair:
                highest_pair = val
    
    return highest_pair
    
def two_pairs(hand_tups):
    numeric_vals = [tup[0] for tup in hand_tups]
    
    pair_count = 0
    
    for val in numeric_vals:
        if numeric_vals.count(val) >= 2:
            
            while val in numeric_vals: # remove all of same value
                numeric_vals.remove(val)
                
            pair_count += 1
    
    if pair_count >= 2:
        return 1
    else:
        return 0
    
def three_of_a_kind(hand_tups):
    numeric_vals = [tup[0] for tup in hand_tups]
    
    for val in numeric_vals:
        if numeric_vals.count(val) >= 3:
            return val
    
    return 0

In [50]:
#teesting
hand_tups = [(5, 'C'), (5, 'D'), (2, 'H'), (2, 'C'), (2, 'D')]

In [15]:
##
high_card(hand_tups)

5

In [16]:
one_pair(hand_tups)

5

In [18]:
two_pairs(hand_tups)

1

In [19]:
three_of_a_kind(hand_tups)

2

In [33]:
hand_tups_2 = [(5, 'C'), (5, 'D'), (5, 'H'), (3, 'C'), (2, 'D')]

In [34]:
two_pairs(hand_tups_2)

0

In [21]:
def straight(hand_tups):
    numeric_vals = [tup[0] for tup in hand_tups]
    
    bottom_val = min(numeric_vals)
    
    if bottom_val+1 in numeric_vals \
        and bottom_val+2 in numeric_vals \
        and bottom_val+3 in numeric_vals \
        and bottom_val+4 in numeric_vals:
    
        return bottom_val
    
    return 0

In [52]:
straight(hand_tups)

0

In [24]:
def flush(hand_tups):
    
    suites = [tup[1] for tup in hand_tups]
    
    if suites[0] == suites[1] and suites[1] == suites[2] \
        and suites[2] == suites[3] and suites[3] == suites[4]:
        
        return 1
    
    return 0

In [51]:
## testing
flush(hand_tups)

0

In [35]:
def full_house(hand_tups):
    
    numeric_vals = [tup[0] for tup in hand_tups]
    
    if three_of_a_kind(hand_tups) > 0:
        triple = three_of_a_kind(hand_tups)
        
        while triple in numeric_vals:
            numeric_vals.remove(triple)
            
        if numeric_vals[0] == numeric_vals[1]:
            return 1
    
    return 0

In [53]:
## testing
full_house(hand_tups)

1

In [25]:
def four_of_a_kind(hand_tups):
    numeric_vals = [tup[0] for tup in hand_tups]
    
    for val in numeric_vals:
        if numeric_vals.count(val) == 4:
            return val
    
    return 0

In [26]:
def straight_flush(hand_tups):
    if flush(hand_tups) > 0 and straight(hand_tups) > 0:
        return flush(hand_tups)
    else:
        return 0

In [27]:
def royal_flush(hand_tups):
    if straight_flush(hand_tups) == 10:
        return 1
    else:
        return 0

In [44]:
def rank_vals(hand_tups):
    """
    return a list of values for
    [royal_flush, straight_flush, four_of_a_kind, full_house,
    flush, straight, three_of_a_kind, two_pairs, one_pair, 
    high_card]

    """
    ranks = []
    
    for rank in [royal_flush, straight_flush, four_of_a_kind, full_house,
        flush, straight, three_of_a_kind, two_pairs, one_pair, 
        high_card]:
        
        ranks.append(rank(hand_tups))
        
    return ranks

In [45]:
## testing
rank_vals(hand_tups)

[0, 0, 0, 1, 0, 0, 2, 1, 5, 5]

In [54]:
def compare(hand_tups_1, hand_tups_2):
    """
    returns 1 if the first hand wins,
        -1 if the second hand wins, 
        0 if both are tied exactly
    
    """
    ranks_1 = rank_vals(hand_tups_1)
    ranks_2 = rank_vals(hand_tups_2)
    
    for i in range(len(ranks_1)):
        
        if ranks_1[i] > ranks_2[i]:
            return 1
        elif ranks_1[i] < ranks_2[i]:
            return -1
        else: # if ranks_1[i] == ranks_2[i]:
            continue
        
    # if all ranks are equal
    return 0

In [55]:
## tinkering
def return_tuple():
    return 1, 2

a, b = return_tuple()

print(a)
print(b)

1
2


# Summary of functions
* card_tuple_from_str(card_str)
    * outputs a tuple of a numeric value and a suite char
* read_hand_string(line)
    * returns two lists of card tuples, for player 1 and 2
    * 
* Rank functions
    * high_card(hand_tups)
    * one_pair(hand_tups)
    * ...
    * royal_flush(hand_tups)
* rank_val(hand_tups)
    * returns a list of rank vals from royal_flush to high_card
* compare(hand_tups_1, hand_tups_2)
    * returns 1 if hand_tups_1 > hand_tups_2, 
        -1 if hand_tups_1 < hand_tups_2,
        0 if equal
* 

In [63]:
## solution code
player_1_win_count = 0

with open('p054_poker.txt') as file:
    
    for line in file:
        hand_tups_1, hand_tups_2 = read_hand_string(line)
        
#         print(hand_tups_1) # error catching
            # found an error in card_tuple_from_str
        
        if compare(hand_tups_1, hand_tups_2) > 0:
            player_1_win_count += 1
            
print(player_1_win_count)

376


# Final answer
Player 1 wins 376 games

Time elapsed:
~6:30 - 7:42

It took me a while just to work up the motivation to start coding. I envisioned that the code wouldn't be difficult, but perhaps just a bit tedious...